Add section about storage management (#11637)
supabase/supabase: apps/docs/components/MDX/storage_management.mdx

> If you repeatedly upload profile photos, you will observe that all uploads will accumulate
> in the `avatars` bucket because of their random names with only the latest being referenced
> from `public.profiles` and the others getting orphaned.
> 
> In order to automatically remove obsolete storage objects, we can extend the database
> triggers. It is important to note that it is not sufficient to delete the objects from the
> `storage.objects` table because that would orphan and leak the actual storage objects in
> the S3 backend. Instead, we'll need to invoke the storage API itself from within Postgres
> via the `http` extension. So, enable that for the `extensions` schema in the
> Database/Extensions dashboard.
> 
> Then, define the following SQL functions in the SQL Editor which allow to easily delete
> storage objects via the API:

Reviewer: ```suggestion
If you upload additional profile photos, they'll accumulate
in the `avatars` bucket because of their random names with only the latest being referenced
from `public.profiles` and the older versions getting orphaned.

To automatically remove obsolete storage objects, extend the database
triggers. Note that it is not sufficient to delete the objects from the
`storage.objects` table because that would orphan and leak the actual storage objects in
the S3 backend. Instead, invoke the storage API within Postgres via the `http` extension. 

Enable the [http extension for the `extensions` schema](https://app.supabase.com/project/_/database/extensions) in the Dashboard.
Then, define the following SQL functions in the SQL Editor to delete
storage objects via the API:
```

```SQL
create or replace function delete_storage_object(bucket text, object text, out status int, out content varchar)
returns record
language 'plpgsql'
security definer
as $$
declare
  project_url varchar := '<YOURPROJECTURL>';
  service_role_key varchar := '<YOURSERVICEROLEKEY>'; --  full access needed
  url varchar := project_url||'/storage/v1/object/'||bucket||'/'||object;
begin
  select
      into status, content
           result.status::int, result.content::varchar
      FROM extensions.http((
    'DELETE',
    url,
    ARRAY[extensions.http_header('authorization','Bearer '||service_role_key)],
    NULL,
    NULL)::extensions.http_request) as result;
end;
$$;

create or replace function delete_avatar(avatar_url text, out status int, out content varchar)
returns record
language 'plpgsql'
security definer
as $$
begin
  select
      into status, content
           result.status, result.content
      from public.delete_storage_object('avatars', avatar_url) as result;
end;
$$;

```

> The next step is to add a trigger that removes any obsolete avatar whenever the
> profile is updated or deleted:

Reviewer: ```suggestion
Next, add a trigger that removes any obsolete avatar whenever the
profile is updated or deleted:
```

```SQL
create or replace function delete_old_avatar()
returns trigger
language 'plpgsql'
security definer
as $$
declare
  status int;
  content varchar;
begin
  if coalesce(old.avatar_url, '') <> ''
      and (tg_op = 'DELETE' or (old.avatar_url <> new.avatar_url)) then
    select
      into status, content
      result.status, result.content
      from public.delete_avatar(old.avatar_url) as result;
    if status <> 200 then
      raise warning 'Could not delete avatar: % %', status, content;
    end if;
  end if;
  if tg_op = 'DELETE' then
    return old;
  end if;
  return new;
end;
$$;

create trigger before_profile_changes
  before update of avatar_url or delete on public.profiles
  for each row execute function public.delete_old_avatar();

```

> The final step is to delete the `public.profile` row ahead of time when a user
> is deleted. If this step is omitted, you won't be able to delete users without
> first manually deleting the avatar image owned by them.

Reviewer: ```suggestion
Finally, delete the `public.profile` row before a user is deleted.
If this step is omitted, you won't be able to delete users without
first manually deleting their avatar image.
```
